synapse/storage/databases/main/roommember.py

```diff
@@ -881,7 +906,7 @@ async def _get_joined_hosts(
         # `get_joined_hosts` is called with the "current" state group for the
         # room, and so consecutive calls will be for consecutive state groups
         # which point to the previous state group.
-        cache = await self._get_joined_hosts_cache(room_id)
+        cache = await self._get_joined_hosts_cache(room_id)  # type: ignore[misc]
```

Comment:
AFAICS `_get_joined_hosts_cache` is defined within this class. What's going on here?

Reply:
That's really weird. I guess there is some `@cached` weirdness that makes this a false positive?

Reply:
`_get_joined_hosts_cache` is declared as:
```py
    @cached(max_entries=10000)
    def _get_joined_hosts_cache(self, room_id: str) -> "_JoinedHostsCache":
```
which is not async. And `@cached`'s type hints aren't too accurate (maybe they can be annotated now that we have Concatenate? be warned that it's a bit of a rabbit hole)

Reply:
I noticed that it's not async, but that made me wonder how the quoted line actually works at runtime. (I'm assuming it's some kind of error to `await x` where `x` isn't `Awaitable`?)

I wondered if it the caching mechanisms were somehow returning a Deferred or another awaitable behind the scenes but... I didn't want to go full rabbit.

Reply:
The caching mechanism does indeed return a `Deferred`, always.

Reply:
Given that, I'd be tempted to change @cached's type so that it always returns `Awaitable[T]` instead of `T`.... but that's a) probably a rabbit hole and b) not one for us to deal with here!